### ISSUE TITLE
多卡运行时，显存负载不均衡，且随着卡的数量增加，第一张卡的显存占用会变大

### ISSUE THREAD

**Issue by cdd1993** [dbiir/UER-py#246]
![gpu_2](https://user-images.githubusercontent.com/28085921/147653495-d8d3261d-83ad-4079-b337-47b3cdc0bd40.png)
![gpu_7](https://user-images.githubusercontent.com/28085921/147653497-9dca6b63-2a34-4644-8790-32c56a37a92f.png)
发现随着训练用的GPU的数量增加，第一张显卡的显存不均衡情况会更加明显，请问作者开源的时候有这个情况吗，是distributed dataparallel哪里参数有问题吗。

**Comment by Abigial61** [comment]
好像我也遇到了类似的问题

**Comment by P01son6415** [comment]
finetune使用的DataParallel会把计算结果放在第一张卡中，所以第一张卡显存高
如果要用distributed dataparallel，可以用pretrain

**Comment by cdd1993** [comment]
> finetune使用的DataParallel会把计算结果放在第一张卡中，所以第一张卡显存高 如果要用distributed dataparallel，可以用pretrain

我这个结果是使用的pretrain的呀

**Comment by P01son6415** [comment]
> > finetune使用的DataParallel会把计算结果放在第一张卡中，所以第一张卡显存高 如果要用distributed dataparallel，可以用pretrain
> 
> 我这个结果是使用的pretrain的呀

可以发一下命令和参数

**Comment by cdd1993** [comment]
> > > finetune使用的DataParallel会把计算结果放在第一张卡中，所以第一张卡显存高 如果要用distributed dataparallel，可以用pretrain
> > 
> > 
> > 我这个结果是使用的pretrain的呀
> 
> 可以发一下命令和参数

python pretrain.py --dataset_path ${model_name}_pretrain_10000.pt \
                     --tokenizer bpe \
                     --merges_path huggingface_pretrained_models/${model_name}/merges.txt \
                     --total_steps ${total_steps} \
                     --save_checkpoint_steps 1000 \
                     --config_path models/${model_type}/${model_name}_config.json \
                      --vocab_path huggingface_pretrained_models/${model_name}/vocab.txt \
                      --pretrained_model_path models/${model_type}/${model_name}.bin \
                      --output_model_path models/${model_type}/tweet_${model_type}_${model_size}_model.bin \
                      --world_size 7 --gpu_ranks 0 1 2 3 4 5 6 --learning_rate 2e-5 \
                      --batch_size ${batch_size} \
                      --embedding word_pos_seg --encoder transformer --mask fully_visible --target mlm > exp_log/m_${model_name}_bs_${batch_size}_ts_${total_steps}_multi_gpu.log 2>&1    非常感谢

**Comment by Abigial61** [comment]
> 

你好，我想请问一下我的pretrain好像会报错 不存在bpe的tokenizer 以及不存在--merges_path这个参数，请问你是怎么解决的呢？

**Comment by P01son6415** [comment]
> > 
> 
> 你好，我想请问一下我的pretrain好像会报错 不存在bpe的tokenizer 以及不存在--merges_path这个参数，请问你是怎么解决的呢？

UER中没有这两个参数，预训练命令可以参考 https://github.com/dbiir/UER-py/wiki/Pretraining-model-examples

**Comment by PALIN2018** [comment]
> > 你好，我想请问一下我的pretrain好像会报错 不存在bpe的tokenizer 以及不存在--merges_path这个参数，请问你是怎么解决的呢？
> 
> UER中没有这两个参数，预训练命令可以参考 https://github.com/dbiir/UER-py/wiki/Pretraining-model-examples

#241 最新版的UER提供了对bpe的支持，以及使用--merges_path。

**Comment by AQA6666** [comment]
> > > finetune使用的DataParallel会把计算结果放在第一张卡中，所以第一张卡显存高 如果要用distributed dataparallel，可以用pretrain
> > 
> > 
> > 我这个结果是使用的pretrain的呀
> 
> 可以发一下命令和参数
我这边预训练使用的预训练命令与你们发布的T5模型基本一样，也出现了相同的问题，0号卡显存占用37G，其余占用23G，且功率只占用了277W~355W/400W。请问你们的多卡并行是怎么实现的，是用的Pytorch的DataParallel还是DistributedDataParallel？据我所知前者更容易出现负载不均衡的问题。

**Comment by cdd1993** [comment]
有人可以解决一下这个问题吗，这个对多卡训练影响很大啊。。。。
